Adding more documentation for upsampling with replacement and error m…

CooperData opened the pull request:
…essage in case replacement is set to False

- [X] closes #27451
- [x] tests added / passed
- [X] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Commit:
Adding more documentation for upsampling with replacement and error message in case replacement is set to False

pep8speaks commented:
Hello @CooperData! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/generic.py`](https://github.com/pandas-dev/pandas/blob/9d26f887a646acd9d06221e048668c40b205f0c0/pandas/core/generic.py):

> [Line 5059:89](https://github.com/pandas-dev/pandas/blob/9d26f887a646acd9d06221e048668c40b205f0c0/pandas/core/generic.py#L5059): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 88 characters)

Commit:
fixing number of characters in line

jreback commented on pandas/core/generic.py:
```diff
@@ -4964,6 +4964,21 @@ def sample(
 
+
+        An upsample sample of the ``DataFrame`` with replacement:
+        Note that `replace` parameter has to be `True` for `frac` parameter > 1.
```
do we have this in the Notes section as well (about replace=True for frac > 1)

CooperData replied:
I added the Notes session

jreback commented on pandas/tests/generic/test_generic.py:
```diff
@@ -424,6 +434,14 @@ def test_sample(self):
+            "Replace has to be set to `True` when upsampling the population `frac` > 1"
+        )
+        with pytest.raises(TypeError, match=msg):
+            hash(df.sample(frac=2, replace=False))
```
don't need the hash

jreback commented on pandas/tests/generic/test_generic.py:
```diff
@@ -424,6 +434,14 @@ def test_sample(self):
         self._compare(o.sample(n=1, axis=0, weights=weights_with_None), o.iloc[5:6])
 
+    def test_sample_upsampling_without_replacement(self):
+        df = pd.DataFrame({"A": list("abc")})
```
can you add the issue number as a comment

CooperData replied:
I removed the hash

jreback reviewed (changes requested):
go ahead and add a whatsnew note, bugfix in 1.0 in the numeric section; you can say that a better error message will be raised when frac > 1 and replace=True

Commit:
Addressing the comments: issue number, remove hash, add Notes

Commit:
adding info about improved message on whatsnew

jreback commented:
lgtm. ping on green.

CooperData commented:
@jreback - CI failed but I am not sure what it's causing the issue. I was able to run the tests while I was at the sprint. But dont know if I did anything that is creating the issue.

jreback commented:
see https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=20307&view=logs&jobId=d6a36677-9147-59f9-684d-001fc5c44c5e

looks like just some extra white space 

fix and push up if u can

Commit:
setting message and test to the same info. Adding dot to the end of the message

CooperData commented:
I added a dot at the end of the raised error as per other errors I saw. I also changed the test msg checked to match the error raised so it's 2 lines as well. I was not able to understand what exactly the error was.

jreback commented:
thanks @CooperData

Commit:
Removing whitespaces from line with no text

Commit:
trying to remove problems with notes

Commit:
removing empty line with spaces

Commit:
adding a dot at the end to see if that would solve the issue

Commit:
double line break found and fixed

CooperData commented:
@jreback, I am going through the messages on the build for the last push, and it seems it's failing because the function inside that is supposed to give the error message is failing. In other words, the desired behavior is to make the test fail but compare the error message, not to make the test fail itself.

Commit:
Comparing to ValueError instead of TypeError

CooperData commented:
The failures seem associated with something else that I did not change: [](https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=20397&view=logs&jobId=5ddd427e-1f95-5998-4195-7ffbeb9b114d&taskId=607906cd-31fb-52fc-bb5f-740f55a1afa6&lineStart=150&lineEnd=186&colStart=1&colEnd=126)

jreback commented:
thanks @CooperData !
